binder/binder.cpp

```diff
@@ -0,0 +1,1150 @@
+                    QMetaObject::invokeMethod(staticProvider, "engineOff", Qt::QueuedConnection);
+                }
+            }
+            }
+            break;
+        case GNSS_SV_STATUS_CB:
+            {
+            const GnssSvStatus* svStatus = geoclue_binder_gnss_decode_struct
+                (GnssSvStatus, &reader);
+
+            QList<SatelliteInfo> satellites;
+            QList<int> usedPrns;
+
+            for (int i = 0; i < svStatus->numSvs; ++i) {
+                SatelliteInfo satInfo;
+                GnssSvInfo svInfo = svStatus->gnssSvList.data[i];
+                satInfo.setPrn(svInfo.svid);
+                satInfo.setSnr(svInfo.cN0Dbhz);
+                satInfo.setElevation(svInfo.elevationDegrees);
+                satInfo.setAzimuth(svInfo.azimuthDegrees);
+                satellites.append(satInfo);
+
+                if (svInfo.svFlag & HYBRIS_GNSS_SV_FLAGS_USED_IN_FIX)
+                    usedPrns.append(svInfo.svid);
+            }
+
+            QMetaObject::invokeMethod(staticProvider, "setSatellite", Qt::QueuedConnection,
+                                      Q_ARG(QList<SatelliteInfo>, satellites),
+                                      Q_ARG(QList<int>, usedPrns));
+            }
+            break;
+        case GNSS_NMEA_CB:
+            {
+            gint64 timestamp;
+            if (gbinder_reader_read_int64(&reader, &timestamp)) {
+                char* nmeaData = gbinder_reader_read_hidl_string(&reader);
+                if (nmeaData) {
+                    int length = strlen(nmeaData);
+                    while (length > 0 && isspace(nmeaData[length-1]))
+                        --length;
+
+                    if (length == 0)
+                        break;
+
+                    QByteArray nmea = QByteArray::fromRawData(nmeaData, length);
+
+                    qCDebug(lcGeoclueHybrisNmea) << timestamp << nmea;
+
+                    if (!nmeaChecksumValid(nmea))
+                        break;
+
+                    // truncate checksum and * from end of sentence
+                    nmea.truncate(nmea.length()-3);
+
+                    if (nmea.startsWith("$GPRMC"))
+                        parseRmc(nmea);
+                }
+            }
+            }
+            break;
+        case GNSS_SET_CAPABILITIES_CB:
+            {
+            guint32 capabilities;
+            if (gbinder_reader_read_uint32(&reader, &capabilities)) {
+                qCDebug(lcGeoclueHybris) << "capabilities" << showbase << hex << capabilities;
+            }
+            }
+            break;
+        case GNSS_ACQUIRE_WAKELOCK_CB:
+        case GNSS_RELEASE_WAKELOCK_CB:
+            break;
+        case GNSS_REQUEST_TIME_CB:
+            qCDebug(lcGeoclueHybris) << "GNSS request UTC time";
+            QMetaObject::invokeMethod(staticProvider, "injectUtcTime", Qt::QueuedConnection);
+            break;
+        case GNSS_SET_SYSTEM_INFO_CB:
```

Comment:
ooi what is this intended for?

Reply:
Currently system info callback only provides "year in which the last update was made to the underlying hardware/firmware used to capture GNSS signals" which is not very relevant for us. We could log it but otherwise we don't have any use for it. The year is used in Android CTS tests to decide which features are required.